File: Tester/FormMain.vb
Imports CoreAudio
Imports System.Threading

Public Class FormMain
    Private timer As Timer

    Private tabLevel As Integer

    Private lines As New List(Of Part)
    Private ctrls As New List(Of Part)

    Private Sub FormMain_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        timer = New Timer(Sub() Me.Invoke(New MethodInvoker(Sub() EnumDevices(EDataFlow.eAll))), Nothing, 10, Timeout.Infinite)

        AddHandler Me.FormClosing, Sub() timer.Dispose()
    End Sub

    Private Sub EnumDevices(flow As EDataFlow)
        Dim mmde As MMDeviceEnumerator = New MMDeviceEnumerator()
        Dim devCol As MMDeviceCollection = mmde.EnumerateAudioEndPoints(flow, DEVICE_STATE.DEVICE_STATE_ACTIVE)
        Dim result As String = ""

        TextBoxOutput.Text = ""

        For d As Integer = 0 To devCol.Count - 1
            Dim dev As MMDevice = devCol(d)
            'Dim epvl As AudioEndpointVolume = dev.AudioEndpointVolume

#If DEBUG Then
            Static isFirstTime As Boolean = True
            If Not isFirstTime Then
                isFirstTime = True
                ' Debug.WriteLine(dev.Properties(PKEY.PKEY_DeviceInterface_FriendlyName))
                For p = 0 To dev.Properties.Count - 1
                    Dim prop As PropertyStoreProperty = dev.Properties(p)
                    Dim valType As String = prop.Value.GetType().ToString()
                    If valType.Contains(".") Then valType = valType.Split(".").Last()
                    Debug.Write($"{prop.Key.fmtid}, {prop.Key.pid} ({valType}): ")

                    Select Case prop.Value.GetType()
                        Case GetType(Byte())
                            For Each b As Byte In CType(prop.Value, Byte())
                                Debug.Write($"{b.ToString("X2")} ")
                            Next
                        Case Else
                            Debug.Write(prop.Value.ToString())
                    End Select
                    Debug.WriteLine("")
                Next
            End If
#End If

            Dim deviceTopology As DeviceTopology = dev.DeviceTopology
            Dim connectorEndpoint As Connector = deviceTopology.GetConnector(0)
            Dim connectorDevice As Connector = connectorEndpoint.GetConnectedTo
            Dim part As Part = connectorDevice.GetPart

            tabLevel = 0
            result += WalkTreeBackwardsFromPart(part)

            result += EnumSessions(dev)

            result += $"{StrDup(32, "-")}{Environment.NewLine}{Environment.NewLine}"
        Next

        TextBoxOutput.Text = result
        TextBoxOutput.SelectionStart = 0
        TextBoxOutput.ScrollToCaret()
    End Sub

    Private Function EnumSessions(dev As MMDevice)
        tabLevel += 1
        Dim tabs As String = StrDup(tabLevel, vbTab)
        Dim result As String = ""

        result += $"{tabs}SESSIONS{Environment.NewLine}"
        Dim sc As SessionCollection = dev.AudioSessionManager2.Sessions
        For i As Integer = 0 To sc.Count - 1
            Dim asc As AudioSessionControl2 = sc.Item(i)
            Dim sName As String = asc.DisplayName
            If sName.Contains(",") Then sName = sName.Split(",")(0)
            If sName = "" Then sName = "<No Name>"
            result += $"{tabs}{sName} [{asc.GetProcessID} | {asc.State}]:{Environment.NewLine}"
            result += $"{tabs}{vbTab}Volume: {asc.SimpleAudioVolume.MasterVolume:F2}dB{Environment.NewLine}"
            result += $"{tabs}{vbTab}Mute: {If(Not asc.SimpleAudioVolume.Mute, "Not ", "")}Muted{Environment.NewLine}{Environment.NewLine}"
            asc.Dispose()
        Next

        Return result
    End Function

    Private Sub GetLines(part As Part)
        If part.GetPartType = PartType.Connector Then lines.Add(part)

        Dim plIn = part.EnumPartsIncoming
        If plIn IsNot Nothing Then
            For i As Integer = 0 To plIn.GetCount - 1
                Dim iPart = plIn.GetPart(i)
                GetLines(iPart)
            Next
        End If
    End Sub

    Private Function WalkTreeBackwardsFromPart(part As Part) As String
        Dim tabs As String = StrDup(tabLevel, vbTab)
        Dim result As String = ""
        result += $"{tabs}{part.GetSubTypeName}: {If(part.GetName = "", "(Unnamed)", part.GetName)} ({part.GetPartType}){Environment.NewLine}"
        'If part.GetSubTypeName = "UNDEFINED" Then Stop

        If part.GetPartType = PartType.Connector Then tabLevel += 1
        Select Case part.GetSubType
            Case KSNODETYPE.VOLUME
                Dim avl As AudioVolumeLevel = part.AudioVolumeLevel
                If avl IsNot Nothing Then
                    For c As Integer = 0 To avl.GetChannelCount - 1
                        Dim lr As AudioVolumeLevel.LevelRange = avl.GetLevelRange(c)
                        result += $"{vbTab} {tabs}Volume for channel {c} is {avl.GetLevel(c):F2}dB (range is {lr.minLevel:F2}dB to {lr.maxLevel:F2}dB in increments of {lr.stepping:F2}dB){Environment.NewLine}"
                    Next
                End If
            Case KSNODETYPE.SPEAKER
                result += $"{tabs}Incoming Parts:{Environment.NewLine}"
            Case KSNODETYPE.MUTE
                Dim am As AudioMute = part.AudioMute
                If am IsNot Nothing Then result += $"{tabs}{vbTab}{If(Not am.Mute, "Not ", "")}Muted{Environment.NewLine}"
            Case KSNODETYPE.PEAKMETER
                Dim apm As AudioPeakMeter = part.AudioPeakMeter
                If apm IsNot Nothing Then
                    For c As Integer = 0 To apm.GetChannelCount - 1
                        result += $"{vbTab}{tabs}Level for channel {c} is {apm.GetLevel(c):F2}dB{Environment.NewLine}"
                    Next
                End If
            Case KSNODETYPE.LOUDNESS
                Dim al As AudioLoudness = part.AudioLoudness
                If al IsNot Nothing Then result += $"{tabs}{vbTab}{If(Not al.Enabled, "Not ", "")}Enabled{Environment.NewLine}"
            Case Else
                If part.GetPartType = PartType.Connector Then
                    result += $"{vbTab}{tabs}I/O Jack{Environment.NewLine}"
                Else
                    result += $"{vbTab}{tabs}Undefined Part: {part.GetSubTypeName}{Environment.NewLine}"
                End If
        End Select

        Dim plIn As PartsList = part.EnumPartsIncoming
        If plIn Is Nothing Then
            result += vbCrLf
        Else
            If plIn.GetCount = 0 Then
                result += vbCrLf
            Else
                For i As Integer = 0 To plIn.GetCount - 1
                    Dim iPart = plIn.GetPart(i)
                    result += WalkTreeBackwardsFromPart(iPart)
                Next
            End If
        End If

        If part.GetPartType = PartType.Connector Then tabLevel -= 1
        part.Dispose()

        Return result
    End Function

    Private Sub ButtonClose_Click(sender As Object, e As EventArgs) Handles ButtonClose.Click
        timer.Change(Timeout.Infinite, Timeout.Infinite)
        Me.Close()
    End Sub

    Private Sub ButtonRefresh_Click(sender As Object, e As EventArgs) Handles ButtonRefresh.Click
        timer.Change(10, Timeout.Infinite)
    End Sub

    Private Sub CheckBoxAutoRefresh_CheckedChanged(sender As Object, e As EventArgs) Handles CheckBoxAutoRefresh.CheckedChanged
        If CheckBoxAutoRefresh.Checked Then
            ButtonRefresh.Enabled = False
            timer.Change(10, 2000)
        Else
            ButtonRefresh.Enabled = True
            timer.Change(Timeout.Infinite, Timeout.Infinite)
        End If
    End Sub

    Private Sub ButtonSave_Click(sender As Object, e As EventArgs) Handles ButtonSave.Click
        Using dlg As New SaveFileDialog()
            With dlg
                .FileName = "CoreAudioNET_Report.txt"
                .Filter = "Text Files|*.txt"
                .InitialDirectory = My.Computer.FileSystem.SpecialDirectories.Desktop
                .Title = "Save CoreAudioNET Report"
                If .ShowDialog(Me) = Windows.Forms.DialogResult.OK Then
                    IO.File.WriteAllText(.FileName, TextBoxOutput.Text)
                End If
            End With
        End Using
    End Sub
End Class

File: samples/CoreAudioConsole.Framework.Discover/FormMain.cs
using CoreAudio;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace CoreAudioConsole.Framework.Discover.Tester {
    public partial class FomMain : Form {
        private readonly string nl = @"\par ";// + Environment.NewLine;

        private System.Threading.Timer timer;
        private int tabLevel;

        private readonly List<Part> lines;
        private readonly List<Part> ctrls;

        private bool isFirstTime;

        private Point rtbScrollPos = new Point();
        private const string rtf = @"{\rtf1\ansi {\colortbl;" +
                                    @"\red245\green245\blue245;" + // cf1 = WhiteSmoke
                                    @"\red000\green192\blue000;" + // cf2 = DarkGreen
                                    @"\red192\green192\blue000;" + // cf3 = DarkYellow
                                    @"\red192\green000\blue192;" + // cf4 = DarkMagenta
                                    @"\red255\green000\blue000;" + // cf5 = Red
                                    @"\red180\green100\blue180;" + // cf6 = LightMagenta
                                    @"\red080\green080\blue255;" + // cf7 = Violet
                                    @"\red080\green080\blue080;" + // cf8 = DarkGray
                                    @"}%\par}";

        public FomMain() {
            InitializeComponent();

            //RichTextBoxOutput.Text = $"A\tB";
            //return;

            Load += (object s, EventArgs e) => {
                timer = new System.Threading.Timer((_) => SafeEnum(), null, 10, Timeout.Infinite);
            };
            FormClosing += (object s, FormClosingEventArgs e) => {
                timer.Dispose();
            };

            lines = new List<Part>();
            ctrls = new List<Part>();

            ButtonClose.Click += (object s, EventArgs e) => Close();
            ButtonRefresh.Click += (object s, EventArgs e) => timer.Change(10, Timeout.Infinite);
            ButtonSave.Click += (object s, EventArgs e) => SaveOutput();
            CheckBoxAutoRefresh.CheckedChanged += (object s, EventArgs e) => {
                if(CheckBoxAutoRefresh.Checked) {
                    timer.Change(10, 850);
                    CheckBoxAutoRefresh.BackColor = Color.FromArgb(192, 0, 0);
                } else {
                    timer.Change(0, Timeout.Infinite);
                    CheckBoxAutoRefresh.BackColor = this.BackColor;

                }
            };
            RichTextBoxOutput.Click += (object s, EventArgs e) => {
                RichTextBoxOutput.SetScrollPos(rtbScrollPos);
            };
        }

        string lastResult = "";
        private void SafeEnum() {
            Task.Run(() => {
                string result = EnumDevices(DataFlow.All);
                if(lastResult != result) {
                    lastResult = result;
                    this.Invoke(new MethodInvoker(() => {
                        rtbScrollPos = RichTextBoxOutput.GetScrollPos();
                        RichTextBoxOutput.SuspendDrawing();
                        RichTextBoxOutput.Rtf = rtf.Replace("%", result.ToString());
                        RichTextBoxOutput.SetScrollPos(rtbScrollPos);
                        RichTextBoxOutput.ResumeDrawing();
                    }));
                }
            });
        }

        private string EnumDevices(DataFlow flow) {
            MMDeviceCollection deviceCollection = new MMDeviceEnumerator(Guid.NewGuid()).EnumerateAudioEndPoints(flow, DeviceState.Active);
            StringBuilder sb = new StringBuilder($@"\cf2 {DateTime.Now}{nl}{nl}");
            for(int i = 0; i < deviceCollection.Count; i++) {
                MMDevice dev = deviceCollection[i];

#if DEBUG
                if(!isFirstTime) {
                    isFirstTime = true;
                    for(int j = 0; j < dev.Properties.Count; j++) {
                        PropertyStoreProperty property = dev.Properties[j];
                        string str2 = property.Value.GetType().ToString();
                        if(str2.Contains(".")) str2 = str2.Split('.').Last();
                        Debug.Write(string.Format("{0}, {1} ({2}): ", property.Key.fmtId, property.Key.PId, str2));
                        if(property.Value.GetType() == typeof(byte[])) {
                            byte[] values = (byte[])property.Value;
                            for(int k = 0; k < values.Length; k++) {
                                Debug.Write(string.Format("{0} ", values[k].ToString("X2")));
                            }
                        } else
                            Debug.Write(property.Value.ToString());
                        Debug.WriteLine("");
                    }
                }
#endif

                Part getPart = dev.DeviceTopology.GetConnector(0).ConnectedTo.Part;
                tabLevel = 0;
                sb.Append(WalkTreeBackwardsFromPart(getPart));
                sb.Append(EnumSessions(dev));

                sb.Append($@"\cf8 ");
                sb.Append('—', RichTextBoxOutput.Width / 12);
                sb.Append($"{nl}{nl}");
            }

            return sb.ToString();
        }

        private string EnumSessions(MMDevice dev) {
            string tabs = "";
            for(int i = 0; i < tabLevel + 1; i++) tabs += @"\tab ";

            StringBuilder sb = new StringBuilder($@"\cf4 {tabs}SESSIONS{nl}");
            SessionCollection sessions = dev.AudioSessionManager2.Sessions;
            for(int i = 0; i < sessions.Count; i++) {
                using(AudioSessionControl2 audioSessionControl2 = sessions[i]) {
                    string name = audioSessionControl2.DisplayName;
                    if(name.Contains(',')) name = name.Split(',')[0];
                    if(name == "") name = "<No Name>";
                    sb.Append($@"\cf3 {tabs}{name} [{audioSessionControl2.ProcessID} | {audioSessionControl2.State}]:{nl}");
                    sb.Append($@"\cf6 {tabs}\tab Volume: {audioSessionControl2.SimpleAudioVolume.MasterVolume:F2}dB{nl}");
                    sb.Append($@"\cf6 {tabs}\tab Mute: {(!audioSessionControl2.SimpleAudioVolume.Mute ? "Not " : "")}Muted{nl}{nl}");
                }
            }
            return sb.ToString();
        }

        private void GetLines(Part part) {
            if(part.PartType == PartType.Connector) lines.Add(part);
            PartsList enumPartsIncoming = part.EnumPartsIncoming;
            if(enumPartsIncoming == null) return;
            for(int i = 0; i < enumPartsIncoming.Count; i++) {
                GetLines(enumPartsIncoming.Part(i));
            }
        }

        private string WalkTreeBackwardsFromPart(Part part) {
            string tabs = "";
            for(int i = 0; i < tabLevel; i++) tabs += @"\tab ";

            StringBuilder sb = new StringBuilder($@"\cf1 {tabs}{part.SubTypeName}: \cf3 {(part.Name == "" ? "(Unnamed)" : part.Name)} ({part.PartType}){nl}");
            // If part.GetSubTypeName = "UNDEFINED" Then Debugger.Break();

            if(part.PartType == PartType.Connector) tabLevel += 1;

            switch(part.SubType) {
                case Guid r when r == KsNodeType.Volume: {
                    AudioVolumeLevel audioVolumeLevel = part.AudioVolumeLevel;
                    if(audioVolumeLevel != null) {
                        for(int i = 0; i < audioVolumeLevel.ChannelCount; i++) {
                            PerChannelDbLevel.LevelRange levelRange = audioVolumeLevel.GetLevelRange(i);
                            sb.Append($@"\cf6\tab {tabs}Volume for channel {i} is {audioVolumeLevel.GetLevel(i):F2}dB (range is {levelRange.MinLevel:F2}dB to {levelRange.MaxLevel:F2}dB in increments of {levelRange.Stepping:F2}dB){nl}");
                        }
                    }

                    break;
                }

                case Guid r when r == KsNodeType.Speaker: {
                    //sb.Append($@"\cf6 {tabs}Incoming Parts:{nl}");
                    break;
                }

                case Guid r when r == KsNodeType.Mute: {
                    AudioMute audioMute = part.AudioMute;
                    if(audioMute != null)
                        sb.Append($@"\cf6 {tabs}\tab {(!audioMute.Mute ? "Not " : "")}Muted{nl}");
                    break;
                }

                case Guid r when r == KsNodeType.PeakMeter: {
                    AudioPeakMeter audioPeakMeter = part.AudioPeakMeter;
                    if(audioPeakMeter != null) {
                        int num = checked(audioPeakMeter.ChannelCount - 1);
                        int channel = 0;
                        while(channel <= num) {
                            sb.Append($@"\cf6 {tabs}\tab Level for channel {channel} is {audioPeakMeter.Level(channel)}dB{nl}");
                            checked { ++channel; }
                        }
                    }
                    break;
                }

                case Guid r when r == KsNodeType.Loudness: {
                    var al = part.AudioLoudness;
                    if(al != null)
                        sb.Append($@"\cf6 {tabs}\tab {(!al.Enabled ? "Not " : "")} Enabled{nl}");
                    break;
                }

                default: {
                    if(part.PartType == PartType.Connector)
                        sb.Append($@"\cf6 {tabs}\tab I/O Jack{nl}");
                    else
                        sb.Append($@"\cf6 {tabs}\tab Undefined Part: {part.SubTypeName}{nl}");
                    break;
                }
            }

            var plIn = part.EnumPartsIncoming;
            if(plIn == null || plIn.Count == 0)
                sb.Append(nl);
            else
                for(int i = 0; i < plIn.Count; i++) {
                    var iPart = plIn.Part(i);
                    sb.Append(WalkTreeBackwardsFromPart(iPart));
                }

            if(part.PartType == PartType.Connector) tabLevel -= 1;
            part.Dispose();
            return sb.ToString();
        }

        private void SaveOutput() {
            using(SaveFileDialog sfDlg = new SaveFileDialog()) {
                sfDlg.FileName = "CoreAudioNET_Report.txt";
                sfDlg.Filter = "Text Files|*.txt";
                sfDlg.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Desktop);
                sfDlg.Title = "Save CoreAudioNET Report";
                if(sfDlg.ShowDialog() == DialogResult.OK) {
                    File.WriteAllText(sfDlg.FileName, RichTextBoxOutput.Text);
                }
            }
        }
    }
}